File: 450-delete-node-in-a-bst/delete-node-in-a-bst.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
 struct TreeNode* find_min(struct TreeNode* root)
{
    if(root==NULL){
        return NULL;
    }
    else if(root->left!=NULL){
        return find_min(root->left);
    }
    return root;
}
struct TreeNode* deleteNode(struct TreeNode* root, int x) {
    if(root==NULL){
        return root;
    }
    else if(x > root->val){
        root->right=deleteNode(root->right,x);
    }
    else if(x<root->val)
        root->left=deleteNode(root->left,x);
    else{
        if(root->left==NULL && root->right==NULL){
            free(root);
            return NULL;
        }
        else if(root->left==NULL || root->right==NULL){
            struct TreeNode* temp;
            if(root->left==NULL){
                temp=root->right;
            }
            else{
                temp=root->left;
            }
            free(root);
            return temp;
        }
        else{
            struct TreeNode* temp=find_min(root->right);
            root->val=temp->val;
            root->right=deleteNode(root->right,temp->val);

        }
    }  
    return root;  
}

File: 450-delete-node-in-a-bst/delete-node-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
        if not root:
            return root
        if root.val>key:
            root.left=self.deleteNode(root.left,key) 
        elif root.val<key:
            root.right=self.deleteNode(root.right,key)
        else:
            if not root.left:
                return root.right
            elif not root.right:
                return root.left
            else:
                cur=root.left
                while cur.right:
                    cur=cur.right
                root.val=cur.val
                root.left=self.deleteNode(root.left,root.val)
        return root            